### ISSUE TITLE
About eps=1e-05 parameter in BN layer

### ISSUE THREAD

**Issue by zhangyedi** [teobaluta/NPAQ#13]
When you encode bnn to cnf, it seems that you ignore _eps=1e-05_ in BN w.r.t. the 87th line in "mlp2cnf/src/bnn.cpp":

	87     float c = -(sqrt(bn_var[0][j]) / bn_weight[0][j]) * bn_bias[0][j] + bn_mean[0][j] - lin_bias[0][j];
In pytorch.document about [BatchNorm1d](https://pytorch.org/docs/stable/generated/torch.nn.BatchNorm1d.html#torch.nn.BatchNorm1d):

	class torch.nn.BatchNorm1d(num_features, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True);

So I think, after training with "_npad/models/bnn_", the resulting model you get will have a parameter _eps=1e-05_ by default in the function of the BN layer. Is this because _bn1.eps_ is too small compared with _bn1.running_var_, and will not affect the final verification result?

**Comment by teobaluta** [comment]
Hi @zhangyedi 

Yes, thanks for the observation! The `eps=1e-5` is there for numerical stability and subject to how the learning framework implements the batch normalization function, e.g. TF2 has a different similar parameter called `variance_epsilon` (see [1]). This parameter would be summed with the `bn_var` under the `sqrt` and the whole constant term would be applied a ceil or a floor function to get rid of the floating points soundly. Thus, that small constant term should be taken care of.

That being said, for some models, we could run into this corner case so it is a valid concern. I did not notice this issue in the models I tried. If you'd like, you could submit a PR with this addition, that'd be great :) 

[1] https://www.tensorflow.org/api_docs/python/tf/nn/batch_normalization
